helix-term/src/commands.rs

```diff
@@ -2092,6 +2093,120 @@ fn symbol_picker(cx: &mut Context) {
+                            }
+                        }
+                    },
+                );
+                compositor.push(Box::new(picker))
+            }
+        },
+    )
+}
+
+fn apply_workspace_edit(
+    editor: &mut Editor,
+    offset_encoding: OffsetEncoding,
+    workspace_edit: &lsp::WorkspaceEdit,
+) {
+    let edits_to_transaction = |doc: &Rope, edits: &Vec<&lsp::TextEdit>| {
+        let lsp_pos_to_pos = |lsp_pos| lsp_pos_to_pos(&doc, lsp_pos, offset_encoding).unwrap();
+        let changes = edits.iter().map(|edit| -> Change {
+            log::debug!("text edit: {:?}", edit);
+            // This clone probably could be optimized if Picker::new would give T instead of &T
+            let text_replacement = Tendril::from(edit.new_text.clone());
+            (
+                lsp_pos_to_pos(edit.range.start),
+                lsp_pos_to_pos(edit.range.end),
+                Some(text_replacement),
+            )
+        });
+        Transaction::change(doc, changes)
+    };
+
+    if let Some(ref changes) = workspace_edit.changes {
+        log::debug!("workspace changes: {:?}", changes);
+        editor.set_error(String::from("Handling workspace changesis not implemented yet, see https://github.com/helix-editor/helix/issues/183"));
+        return;
+        // Not sure if it works properly, it'll be safer to just panic here to avoid breaking some parts of code on which code actions will be used
+        // TODO: find some example that uses workspace changes, and test it
+        // for (url, edits) in changes.iter() {
+        //     let file_path = url.origin().ascii_serialization();
+        //     let file_path = std::path::PathBuf::from(file_path);
+        //     let file = std::fs::File::open(file_path).unwrap();
+        //     let mut text = Rope::from_reader(file).unwrap();
+        //     let transaction = edits_to_changes(&text, edits);
+        //     transaction.apply(&mut text);
+        // }
+    }
+
+    if let Some(ref document_changes) = workspace_edit.document_changes {
+        match document_changes {
+            lsp::DocumentChanges::Edits(document_edits) => {
+                for document_edit in document_edits {
+                    let (view, doc) = current!(editor);
+                    assert_eq!(doc.url().unwrap(), document_edit.text_document.uri);
+                    let edits = document_edit
+                        .edits
+                        .iter()
+                        .map(|edit| match edit {
+                            lsp::OneOf::Left(text_edit) => text_edit,
+                            lsp::OneOf::Right(annotated_text_edit) => {
+                                &annotated_text_edit.text_edit
+                            }
+                        })
+                        .collect();
+                    let transaction = edits_to_transaction(doc.text(), &edits);
+                    doc.apply(&transaction, view.id);
```

Comment:
Undo doesn't work after applying a change:

```suggestion
                    doc.apply(&transaction, view.id);
                    doc.append_changes_to_history(view.id);
```

Reply:
> doc.append_changes_to_history(view.id);

It did work after pressing ESC, but with this change works even without it. I can't commit changes since the PR have been already merged. Maybe you could create a new PR with this change?

Reply:
Ah, ESC is bound to `normal_mode` command (even inside normal mode) and it has an `append_changes_to_history` call. I see you created a new PR 👍🏾